feat(attach): support attachments on cron triggers

boc-tothefuture opened the pull request:
Closes #360

boc-tothefuture commented on lib/openhab/dsl/rules/automation_rule.rb:
```diff
@@ -145,6 +145,10 @@ def add_attachment(event, inputs)
           attachment = @attachments[trigger_id(inputs)]
           return event unless attachment
 
+          # Some events, like those from cron triggers don't have an event
+          # so an event is created
+          event ||= Struct.new(:attachment).new
+
```
hmm.  it didn't work before I added that..  I will have to look again.

boc-tothefuture commented on lib/openhab/dsl/rules/rule.rb:
```diff
@@ -84,13 +85,44 @@ def re_raise_with_backtrace(error)
+      # Add cron triggers with attachments to rules
+      # @param [Array] cron_attach_triggers cron type triggers with attachments
+      #
+      def process_cron_attach(cron_attach_triggers, config, rule)
```
But it is only cron triggers with attachments - not all cron triggers.  Still think the name should change?

jimtng replied:
Ohh, you're right. Leave the name as is.

boc-tothefuture commented on lib/openhab/dsl/rules/rule.rb:
```diff
@@ -84,13 +85,44 @@ def re_raise_with_backtrace(error)
+      #
+      def process_cron_attach(cron_attach_triggers, config, rule)
+        cron_attach_triggers&.map { |trigger| CronTriggerRule.new(rule_config: config, rule: rule, trigger: trigger) }
+                            &.each { |trigger| add_rule(trigger) }
```
I don't think so? Because the partition is only for cron triggers with attachments.

jimtng replied:
I tried running the cucumber test line 33, and saw two `Adding rule...`  in the log:
```
00:40:26.972 [INFO ] [jsr223.jruby.Access_attachment       ] - 31 40 00 ? * ?
00:40:26.972 [TRACE] [jsr223.jruby.Access_attachment       ] - Creating trigger of type timer.GenericCronTrigger for {"cronExpression"=>"31 40 00 ? * ?"}
00:40:26.973 [TRACE] [jsr223.jruby.Access_attachment       ] - Property 'run' called with args([]) and block(#<Proc:0x66d90920@/Users/jimmy/Projects/openhab-jruby/tmp/openhab/conf/automation/jsr223/ruby/personal/cucumber_test20211201-74581-v58v8z.rb:13>)
00:40:26.977 [TRACE] [jsr223.jruby.Object                  ] - Name: (Access attachment) Triggers: ([#<Java::OrgOpenhabCoreAutomationInternal::TriggerImpl:0x7fd40cd5>]) Run blocks: ([#<struct OpenHAB::DSL::Rules::RuleConfig::Run block=#<Proc:0x66d90920@/Users/jimmy/Projects/openhab-jruby/tmp/openhab/conf/automation/jsr223/ruby/personal/cucumber_test20211201-74581-v58v8z.rb:13>>]) on_start: (false) Trigger Waits: {} Trigger UIDs: 071de2de-58d0-4124-a677-c224ae8f79b7Attachments: {"071de2de-58d0-4124-a677-c224ae8f79b7"=>"qab"} 
00:40:26.978 [TRACE] [jsr223.jruby.Object                  ] - Cron triggers: [#<Java::OrgOpenhabCoreAutomationInternal::TriggerImpl:0x7fd40cd5>] -  Other triggers: []
00:40:26.980 [TRACE] [jsr223.jruby.Object                  ] - Adding rule: #<OpenHAB::DSL::Rules::AutomationRule:0x79847aca @between=#<OpenHAB::DSL::Between::TimeOfDayRangeElement:0x51f57b2e @range_begin=0, @sod=0>..#<OpenHAB::DSL::Between::TimeOfDayRangeElement:0x57a248b1 @range_begin=0, @sod=86399>, @run_context=main, @trigger_delays={}, @run_queue=[#<struct OpenHAB::DSL::Rules::RuleConfig::Run block=#<Proc:0x66d90920@/Users/jimmy/Projects/openhab-jruby/tmp/openhab/conf/automation/jsr223/ruby/personal/cucumber_test20211201-74581-v58v8z.rb:13>>], @guard=#<OpenHAB::DSL::Rules::Guard::Guard:0xdaa2aa5 @only_if=nil, @not_if=nil>, @attachments={"071de2de-58d0-4124-a677-c224ae8f79b7"=>"qab"}>
00:40:26.980 [TRACE] [.automation.internal.RuleEventFactory] - creating ruleEvent of type: RuleAddedEvent
00:40:26.981 [TRACE] [.automation.internal.RuleEventFactory] - creating ruleEvent of type: RuleStatusInfoEvent
00:40:26.981 [TRACE] [.automation.internal.RuleEventFactory] - creating ruleEvent of type: RuleStatusInfoEvent
00:40:26.981 [TRACE] [.automation.internal.RuleEventFactory] - creating ruleEvent of type: RuleStatusInfoEvent
00:40:26.982 [TRACE] [obj.OpenHAB.DSL.Rules.CronTriggerRule] - Created Cron Trigger Rule for org.openhab.core.automation.internal.TriggerImpl@7fd40cd5
00:40:26.982 [TRACE] [jsr223.jruby.Object                  ] - Adding rule: #<OpenHAB::DSL::Rules::CronTriggerRule:0x6704aded @trigger=#<Java::OrgOpenhabCoreAutomationInternal::TriggerImpl:0x7fd40cd5>, @rule=#<OpenHAB::DSL::Rules::AutomationRule:0x79847aca @between=#<OpenHAB::DSL::Between::TimeOfDayRangeElement:0x51f57b2e @range_begin=0, @sod=0>..#<OpenHAB::DSL::Between::TimeOfDayRangeElement:0x57a248b1 @range_begin=0, @sod=86399>, @run_context=main, @trigger_delays={}, @run_queue=[#<struct OpenHAB::DSL::Rules::RuleConfig::Run block=#<Proc:0x66d90920@/Users/jimmy/Projects/openhab-jruby/tmp/openhab/conf/automation/jsr223/ruby/personal/cucumber_test20211201-74581-v58v8z.rb:13>>], @guard=#<OpenHAB::DSL::Rules::Guard::Guard:0xdaa2aa5 @only_if=nil, @not_if=nil>, @attachments={"071de2de-58d0-4124-a677-c224ae8f79b7"=>"qab"}>>
```

jimtng replied:
It seems there could be some kind of DRYing that needs to happen between AutomationRule vs CronTriggerRule

boc-tothefuture replied:
No, that is correct.  You should have two rules.  The overall rule (that may have more than one trigger and may or may not have a cron trigger with an attachment and the cron trigger rule - which creates another rule that invokes execute on that first rule.

boc-tothefuture replied:
Depending on what I learn when creating the directory watch trigger, I may be able to refactor this in the future to remove the creation of the secondary rule.. I will create an issue for that so we don't lose track of it, but I don't want to hold off on merging this for that potential future improvement.

jimtng replied:
> Depending on what I learn when creating the directory watch trigger, I may be able to refactor this in the future to remove the creation of the secondary rule.. I will create an issue for that so we don't lose track of it, but I don't want to hold off on merging this for that potential future improvement.

OK. Ideally I'd like to end up with just one rule but I agree, it can be refactored later.

boc-tothefuture commented:
@jimtng I think I addressed your issues - extra struct for event attachment is now gone and the names for the methods stayed the same based on our discussion and the 'duplicate' rule is explained.

jimtng commented on lib/openhab/dsl/rules/automation_rule.rb:
```diff
@@ -145,6 +146,10 @@ def add_attachment(event, inputs)
           attachment = @attachments[trigger_id(inputs)]
           return event unless attachment
 
+          # Some events, like those from cron triggers don't have an event
+          # so an event is created
+          #event ||= Struct.new(:attachment).new
+
```
Do we want to just remove this or do you have plans for it in the future?

jimtng commented on lib/openhab/dsl/rules/rule.rb:
```diff
@@ -1,7 +1,8 @@
+require_relative 'rule_config'
+require_relative 'automation_rule'
+require_relative 'cron_trigger_rule'
 require 'openhab/dsl/rules/guard'
```
Why stop here?

jimtng commented:
> @jimtng I think I addressed your issues - extra struct for event attachment is now gone and the names for the methods stayed the same based on our discussion and the 'duplicate' rule is explained.

I only noticed two cosmetic issues, otherwise it looks good.

jimtng commented on lib/openhab/dsl/rules/automation_rule.rb:
```diff
@@ -106,6 +106,7 @@ def create_queue(inputs)
         def extract_event(inputs)
           event = inputs&.dig('event')
           unless event
+            puts "Creating new event struct"
```
might want to remove this
